### PR TITLE
Add parquet roundtrip test

### PR DESCRIPTION
**Pull request by mrocklin**
@martindurant here is a simple parametrized test that we might use to track progress and quickly find some issues.  We should also feel free to mark these as xfailed until they work.  What else can you recommend adding here?

### CODE REVIEW

**Comment by martindurant** [comment]
Note on the M8[ns] failures: parquet can only store up to a precision of us, so these are just rounding errors.
The TypeError when passing strings is expected, because we didn't specify how to encode the objects (for raw fastparquet you can 'infer', but this is disabled in dask, because we don't want to force computing the first partition). If the default were utf8, then the bytes version would error.

File: dask/dataframe/io/tests/test_parquet.py [hunk 1]
```diff
@@ -134,3 +134,41 @@ def test_categorical():
...
         df.index.name = 'index'  # defaults to 'index' in this case
         assert assert_eq(df, ddf2)
+
+
+@pytest.mark.parametrize('df,kwargs', [
+    (pd.DataFrame({'x': [3, 2, 1]}), {}),
+    (pd.DataFrame({'x': ['c', 'a', 'b']}), {'object_encoding': 'utf8'}),
+    (pd.DataFrame({'x': ['cc', 'a', 'bbb']}), {'object_encoding': 'utf8'}),
+    (pd.DataFrame({'x': [b'a', b'b', b'c']}), {'object_encoding': 'bytes'}),
+    (pd.DataFrame({'x': pd.Categorical(['a', 'b', 'c'])}), {}),
```
Comment (martindurant) [inline dask/dataframe/io/tests/test_parquet.py]: These are also objects requiring object_encoding. It occurs to me that for the specific case of category labels, we could enable 'infer' after all.

File: dask/dataframe/io/tests/test_parquet.py [hunk 2]
```diff
@@ -134,3 +134,41 @@ def test_categorical():
...
+    (pd.DataFrame({'x': [3, 2, 1]}).astype('M8[us]'), {}),
+    (pd.DataFrame({'x': [3, 2, 1]}).astype('M8[ms]'), {}),
+    (pd.DataFrame({'x': [3, 2, 1]}).astype('uint16'), {}),
+    (pd.DataFrame({'x': [3, 2, 1]}).astype('float32'), {}),
+    (pd.DataFrame({'x': [3, 1, 2]}, index=[3, 2, 1]), {}),
+    (pd.DataFrame({'x': [3, 1, 5]}, index=pd.Index([1, 2, 3], name='foo')), {}),
+    (pd.DataFrame({'x': [1, 2, 3],
+                  'y': [3, 2, 1]}), {}),
+    (pd.DataFrame({'x': [1, 2, 3],
+                  'y': [3, 2, 1]}, columns=['y', 'x']), {}),
```
Comment (martindurant) [inline dask/dataframe/io/tests/test_parquet.py]: Requires https://github.com/dask/dask/pull/1862

**Comment by mrocklin** [comment]
Should we xfail categoricals for now? We currently don't roundtrip these well

File: dask/dataframe/io/tests/test_parquet.py [hunk 3]
```diff
@@ -152,3 +152,42 @@ def test_ordering():
...
+])
+def test_roundtrip(df, kwargs):
+    with tmpdir() as tmp:
+        tmp = str(tmp)
+        if df.index.name is None:
+            df.index.name = 'index'
+        ddf = dd.from_pandas(df, npartitions=2)
+
+        to_parquet(tmp, ddf, **kwargs)
+        ddf2 = read_parquet(tmp, index=df.index.name)
```
Comment (martindurant) [inline dask/dataframe/io/tests/test_parquet.py]: You would need `categories=` to get the dtype right, but you can add `check_categorical=False` to the assert_equal, that I think should check the actual values.

Comment (martindurant) [inline dask/dataframe/io/tests/test_parquet.py]: That's for pd.util.testing.assert_frame_equal anyway - of could explicitly check the columns for equality.

**Pull request merged**